### ISSUE TITLE
point-wise scores

### ISSUE THREAD

**Issue by xinyanghuang7** [CodeGoat24/UnifiedReward#1]
Hello author, thank you for your open source. 

How can I generate point-wise scores for a single image and text? 

Is there any code I can refer to? 

Thank you for your great work!

**Comment by CodeGoat24** [comment]
Thanks for your reminder!!

We will release all pair- and point-wise inference codes tomorrow!

**Comment by xinyanghuang7** [comment]
Thank you very much for your prompt reply. Can't wait to see the relevant content.

**Comment by yuqiliang0422** [comment]
Hello, Thanks for your great work. I have used the inference code to generate point-wise scores in the EvalMuse val dataset. Most of the scores are 3.3, 3.7, and 4.7.  And the SRCC, and PLCC with GT are 0.0xxx. 
This is very strange because you tuned the model in EvalMuse40K.

**Comment by CodeGoat24** [comment]
Thank you for your feedback!

During training, the final score of EvalMuse's data is obtained by calculating the average of the total score list in the dataset.

We have re-examined the point sift scoring results when constructing the image generation DPO preference data and found no anomalies.

Could you provide your evaluation script? We will actively investigate the issue and keep you updated.

**Comment by yuqiliang0422** [comment]
Thanks for your quick reply~, here is my script.

Dataset is here 
https://drive.google.com/file/d/1gIN8hqTGS7yki74RzW8w4LGUPJlozonl/view

from llava.model.builder import load_pretrained_model
from llava.mm_utils import get_model_name_from_path, process_images, tokenizer_image_token
from llava.constants import IMAGE_TOKEN_INDEX, DEFAULT_IMAGE_TOKEN, DEFAULT_IM_START_TOKEN, DEFAULT_IM_END_TOKEN, IGNORE_INDEX
from llava.conversation import conv_templates, SeparatorStyle

from PIL import Image
import requests
import copy
import torch

import sys
import warnings
import os
from datasets import load_dataset
import tqdm
import json


warnings.filterwarnings("ignore")

pretrained = "CodeGoat24/UnifiedReward-7b"

model_name = "llava_qwen"
device = "cuda"
device_map = "auto"

tokenizer, model, image_processor, max_length = load_pretrained_model(pretrained, None, model_name, device_map=device_map)  # Add any other thing you want to pass in llava_model_args

model.eval()

from tqdm import tqdm

# 假设原始的 JSON 文件名为 original.json
input_filename = 'test.json'
output_filename = 'test_pred.json'

# 读取原始的 JSON 文件
with open(input_filename, 'r', encoding='utf-8') as f:
    original_data = json.load(f)

# 创建新的列表，用于存储转换后的数据
new_data = []

# 遍历原始 JSON 数据并按照规则生成新的字典
for value in tqdm(original_data):
    # 构建每个字典
    image_path = value['image']
    prompt= value['prompt']
    # answer_value = f"{int(value['mos'])}"

    # 将构建好的字典加入到新的列表中
    
    
    #image_path = '/extra_room/junlin/EvalMuse-Structure/data/train/images/451f3c4e654d11efb45852fe1602b694.jpg'

    #prompt ='A woman riding a Tyrannosaurus Rex in the deep sea, with a school of fish in the background forming a question mark.' #prompt of the given image


    image = Image.open(image_path).resize((512, 512))

    image_tensor = process_images([image], image_processor, model.config)
    image_tensor = [_image.to(dtype=torch.float16, device=device) for _image in image_tensor]

    conv_template = "qwen_1_5"  # Make sure you use correct chat template for different models

    question = f'<image>\nYou are given a text caption and a generated image based on that caption. Your task is to evaluate this image based on two key criteria:\n1. Alignment with the Caption: Assess how well this image aligns with the provided caption. Consider the accuracy of depicted objects, their relationships, and attributes as described in the caption.\n2. Overall Image Quality: Examine the visual quality of this image, including clarity, detail preservation, color accuracy, and overall aesthetic appeal.\nBased on the above criteria, assign a score from 1 to 5 after \'Final Score:\'.\nYour task is provided as follows:\nText Caption: [{prompt}]'

    conv = copy.deepcopy(conv_templates[conv_template])
    conv.append_message(conv.roles[0], question)
    conv.append_message(conv.roles[1], None)
    prompt_question = conv.get_prompt()

    input_ids = tokenizer_image_token(prompt_question, tokenizer, IMAGE_TOKEN_INDEX, return_tensors="pt").unsqueeze(0).to(device)
    image_sizes = [image.size]

    cont = model.generate(
        input_ids,
        images=image_tensor,
        image_sizes=image_sizes,
        do_sample=False,
        temperature=0,
        max_new_tokens=4096
    )
    text_outputs = tokenizer.batch_decode(cont, skip_special_tokens=True)

    
        
    new_data.append({
        "image": image_path,
        "prompt": prompt,
        "answer": text_outputs[0]
    })
    print(new_data[-1])

# 将新的数据写入到文件
with open(output_filename, 'w', encoding='utf-8') as f:
    json.dump(new_data, f, ensure_ascii=False, indent=4)

print(f"新 JSON 文件已经保存为 {output_filename}")

**Comment by CodeGoat24** [comment]
```
"<image>\nYou are given a text caption and a generated image based on that caption. Your task is to evaluate this image based on two key criteria:\n1. Alignment with the Caption: Assess how well this image aligns with the provided caption. Consider the accuracy of depicted objects, their relationships, and attributes as described in the caption.\n2. Overall Image Quality: Examine the visual quality of this image, including clarity, detail preservation, color accuracy, and overall aesthetic appeal.\nExtract key elements from the provided text caption, evaluate their presence in the generated image using the format: \'element (type): value\' (where value=0 means not generated, and value=1 means generated), and assign a score from 1 to 5 after \'Final Score:\'.\nYour task is provided as follows:\nText Caption: [{prompt}]"
This is the input prompt we used during training. Below is an example output using test data from [EvalMuse](https://huggingface.co/datasets/DY-Evalab/EvalMuse/tree/main) 'test.json' from my recent run:
```
image caption:
```
attack of the pig giants, giant pigs, illustration, children\'s book, fictional drawing, bright colors, fantasy world, minimalist
```
Image

output：
```
pig giants (animal): 0 
giant pigs (animal): 1 
attack (activity): 0 
illustration (attribute): 1 
children's book (object): 1 
fictional drawing (attribute): 1 
bright colors (attribute): 1 
fantasy world (other): 1 
minimalist (attribute): 1 
Final Score: 2.3
```
You can try using the input prompt we provided, and If have any other questions, feel free to contact us.

**Comment by xinyanghuang7** [comment]
Thank you for your positive reply!

I had a similar problem.

From the examples you provide, it seems that the model can directly deal with the question of whether attributes exist or not.

But from the prompt you provided, it seems that there are no special constraints on the output of the model.

So how does the script capture the output of the model, the structured content in the output that you provide?

Is there any additional processing done to capture the relevant part of the model's answer attribute?

Can you provide some code examples?

Thank you very much!!

**Comment by xinyanghuang7** [comment]
This is an insightful approach.

However, if this method is adopted, given that the 'test. json' file from EvalMuse provides certain attributes that need to be pre-evaluated by the model, would there be any potential conflicts between the model's output and these attributes?

Please note that this is merely a minor concern.

**Comment by CodeGoat24** [comment]
Thanks for your feedback！

The last sample above is not from the test set of EvalMuse. 

To our knowledge, there should be no potential issues. We also used this method for point sifting when constructing the image generation preference data. 

If you encounter any problems, please contact us at any time.
